File: backend/app/schemas/trainer_feedback.py
"""API contracts for trainer feedback imports and performance analytics."""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


class RatingColumnMapping(BaseModel):
    column: str = Field(min_length=1, max_length=255)
    max_score: Optional[int] = Field(default=None, ge=1, le=100)


class TrainerFeedbackMapping(BaseModel):
    trainer_column: str = Field(min_length=1, max_length=255)
    submitted_at_column: Optional[str] = Field(default=None, max_length=255)
    feedback_column: Optional[str] = Field(default=None, max_length=255)
    rating_columns: list[RatingColumnMapping] = Field(min_length=1, max_length=30)

    @model_validator(mode="after")
    def unique_rating_columns(self):
        columns = [entry.column for entry in self.rating_columns]
        if len(columns) != len(set(columns)):
            raise ValueError("Each rating column can only be selected once")
        return self


class TrainerFeedbackPreview(BaseModel):
    filename: str
    source_format: str
    worksheet_name: Optional[str] = None
    headers: list[str]
    row_count: int
    sample_rows: list[dict[str, str]]
    suggested_mapping: TrainerFeedbackMapping | dict


class TrainerFeedbackImportResult(BaseModel):
    import_id: UUID
    imported_rows: int
    skipped_rows: int
    duplicate: bool = False


class TrainerAxisScore(BaseModel):
    name: str
    score: float
    response_count: int


class TrainerPerformanceItem(BaseModel):
    rank: int
    trainer_name: str
    response_count: int
    overall_score: Optional[float] = None
    confidence: str
    status: str
    axis_scores: list[TrainerAxisScore]
    feedback_count: int
    recent_feedback: list[str]


class TrainerPerformanceAnalytics(BaseModel):
    total_trainers: int
    total_responses: int
    overall_score: Optional[float] = None
    latest_import_at: Optional[datetime] = None
    trainers: list[TrainerPerformanceItem]

File: src/services/trainerFeedback.ts
import api from './api';

export interface RatingColumnMapping {
  column: string;
  max_score: number | null;
}

export interface TrainerFeedbackMapping {
  trainer_column: string;
  submitted_at_column: string | null;
  feedback_column: string | null;
  rating_columns: RatingColumnMapping[];
}

export interface TrainerFeedbackPreview {
  filename: string;
  source_format: string;
  worksheet_name: string | null;
  headers: string[];
  row_count: number;
  sample_rows: Record<string, string>[];
  suggested_mapping: TrainerFeedbackMapping;
}

export interface TrainerAxisScore {
  name: string;
  score: number;
  response_count: number;
}

export interface TrainerPerformanceItem {
  rank: number;
  trainer_name: string;
  response_count: number;
  overall_score: number | null;
  confidence: string;
  status: string;
  axis_scores: TrainerAxisScore[];
  feedback_count: number;
  recent_feedback: string[];
}

export interface TrainerPerformanceAnalytics {
  total_trainers: number;
  total_responses: number;
  overall_score: number | null;
  latest_import_at: string | null;
  trainers: TrainerPerformanceItem[];
}

const upload = async <T>(url: string, file: File, mapping?: TrainerFeedbackMapping): Promise<T> => {
  const formData = new FormData();
  formData.append('file', file);
  if (mapping) formData.append('mapping_json', JSON.stringify(mapping));
  const response = await api.post<{ data: T }>(url, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return response.data.data;
};

export const previewTrainerFeedbackFile = (file: File) => upload<TrainerFeedbackPreview>('/trainer-feedback/preview', file);

export const importTrainerFeedbackFile = (file: File, mapping: TrainerFeedbackMapping) => upload<{
  import_id: string;
  imported_rows: number;
  skipped_rows: number;
  duplicate: boolean;
}>('/trainer-feedback/imports', file, mapping);

export const getTrainerFeedbackAnalytics = async (): Promise<TrainerPerformanceAnalytics> => {
  const response = await api.get<{ data: TrainerPerformanceAnalytics }>('/trainer-feedback/analytics');
  return response.data.data;
};
